### PR TITLE
DRILL-4580: Support for exporting storage plugin configurations

### PR DESCRIPTION
**Pull request by agirish**
Added an "Export" button to enabled storage plugins to support exporting the configuration into a JSON file. This can be useful to backup storage plugin configurations for Upgrade or other purposes.

### CODE REVIEW

File: exec/java-exec/src/main/resources/rest/storage/list.ftl [hunk 1]
```diff
@@ -38,6 +38,7 @@
               <td style="border:none;">
                 <a class="btn btn-primary" href="/storage/${plugin.getName()}">Update</a>
                 <a class="btn btn-default" onclick="doEnable('${plugin.getName()}', false)">Disable</a>
+                <a class="btn btn-default" href="/storage/${plugin.getName()}/export"">Export</a>
```
[the remark this answers is not in the document]
Comment (Agirish) [reply exec/java-exec/src/main/resources/rest/storage/list.ftl]: I've added it for all plugins. For Enabled it is visible in the Storage Plugin List Page. For Enabled + Disabled, it's also visible in the Update page. 

We can make it visible for Disabled plugins in the main page - but kept it this way to reduce UI clutter. We can also remove the Export button for enabled plugins in the main page. Can you look at the screenshots and see what's preferable?

File: exec/java-exec/src/main/resources/rest/storage/update.ftl [hunk 2]
```diff
@@ -48,6 +48,7 @@
       <#else>
         <a id="enabled" class="btn btn-primary">Enable</a>
       </#if>
+      <a class="btn btn-default" href="/storage/${model.getName()}/export"">Export</a>
```
[the remark this answers is not in the document]
Comment (Agirish) [reply exec/java-exec/src/main/resources/rest/storage/update.ftl]: image

image

[the remark this answers is not in the document]
Comment (chunhui-shi) [reply exec/java-exec/src/main/resources/rest/storage/update.ftl]: +1, do we have documents/commitments to keep these REST APIs stable? If so I think we also need some unit tests testing these APIs

[the remark this answers is not in the document]
Comment (Agirish) [reply exec/java-exec/src/main/resources/rest/storage/update.ftl]: Thanks Chunhui! I am not sure what's needed to qualify these APIs as stable. Can you please clarify? I don't foresee anything which could break them in future releases, as it's pretty generic. 

Also, I couldn't find any sample unit tests for these REST APIs - so I'm not sure how to proceed with it.

[the remark this answers is not in the document]
Comment (chunhui-shi) [reply exec/java-exec/src/main/resources/rest/storage/update.ftl]: Yeah, that is true - this is not covered by any auto-tests. So you could just log a JIRA first and add tests later.

[the remark this answers is not in the document]
Comment (Agirish) [reply exec/java-exec/src/main/resources/rest/storage/update.ftl]: Sure, thanks! I'll do that.

**Comment by arina-ielchiieva** [comment]
@Agirish we might consider adding button to export all enabled plugins in one json file as well, rather then doing one by one. What do you think?

**Comment by Agirish** [comment]
@arina-ielchiieva, I thought about that - but it's a bit more complicated in how i'd like it to be: An "Export All" button" should export each of the storage plugin configurations into JSON files and generate a single compressed folder which contains each of these JSON files individually. 
And an "Import" button which can support importing either a single JSON file or a compressed folder which can create one or more storage plugins. 

I think having individual export button is helpful. We could keep it in the "Update" page (image 2 above). We could remove it from the main storage plugin list page (image 1) if you think it's cluttering the UI. 

In a later PR, we can add support for the "Export All" and "Import" (one or more) buttons. Would that be okay?

**Comment by arina-ielchiieva** [comment]
Sounds fair regarding the enhancement Jira. I think keeping export button even in update mode is also good. Please create Jira for the enhancement for import / export all. Also why you think import all should export each plugin in different file, it can export all in one (similar as we have for bootstrap plugins)?

**Comment by Agirish** [comment]
Filed DRILL-6562 to track this. 

Regarding one vs multiple files, my reasoning was to keep it more human readable and flexible - giving options to selectively import any plugin as needed. So suggested a compressed folder with each of the plugin files. But your approach sounds good. I'm thinking the "Export All" option would be used for All or None scenarios. And the regular per plugin "Export" option being introduced in this PR can support the flexible per plugin import.

**Comment by arina-ielchiieva** [comment]
@kkhatua it can be useful, it you want to back all plugins rather then exporting one by one, the same as with upload. Anyway, it would just an option and up to the user which way to use.

@Agirish thanks for creating the Jira. 
+1 for the current PR.

**Comment by Agirish** [comment]
Thanks all, for the review!

**Pull request closed without merging**